agentlightning/execution/client_server.py

```diff
@@ -196,6 +206,71 @@ async def _execute_runner(
                 else:
                     logger.debug("Runner %s closed LightningStore client", worker_id)
 
+    def _run_with_sigint(
+        self,
+        kind: Literal["runner", "algorithm"],
+        *,
+        runner: RunnerBundle | None = None,
+        algorithm: AlgorithmBundle | None = None,
+        worker_id: int = 0,
+        store: LightningStore,
+        stop_evt: ExecutionEvent,
+        is_main_process: bool,
+    ) -> None:
+        """Run `_execute_runner` or `_execute_algorithm` in a fresh event loop with a SIGINT handler."""
+
+        if kind == "runner":
+            if is_main_process:
+                log_prefix = "Main runner process"
+            else:
+                log_prefix = f"Runner process {worker_id}"
+        else:  # kind == "algorithm"
+            log_prefix = "Algorithm process"
+
+        if kind == "runner":
+            assert runner is not None
+            make_coro = lambda: self._execute_runner(runner, worker_id, store, stop_evt)
+        else:
+            assert algorithm is not None
+            make_coro = lambda: self._execute_algorithm(algorithm, store, stop_evt)
+
+        loop = asyncio.new_event_loop()
+        asyncio.set_event_loop(loop)
+
+        # Explicitly create the main task so the SIGINT handler can do cancel()
+        main_task: asyncio.Task[None] = loop.create_task(make_coro())
+
+        def _sigint_handler(signum: int, frame: object | None) -> None:
+            global SIGINT_SEEN
+            SIGINT_SEEN = True
+            logger.debug("%s received SIGINT; cancelling main task", log_prefix)
+            # Turn SIGINT into cancellation of the main "_execute_*" task so
+            # the runner/algorithm receives asyncio.CancelledError.
+            main_task.cancel()
+
+        signal.signal(signal.SIGINT, _sigint_handler)
```

Comment:
I think this registration is process wide. Let's be extra careful on that.

try: except KeyboardInterrupt might be a better idea if that works.